pkg/service/backup/worker_deduplicate.go

```diff
@@ -0,0 +1,168 @@
+		for _, file := range d.Progress.files {
+			id := sstable.ExtractID(file.Name)
+			ssTablesGroupByID[id] = append(ssTablesGroupByID[id], file.Name)
+		}
+
+		if err := w.basedOnUUIDGenerationAvailability(ctx, d, h, remoteFiles, ssTablesGroupByID); err != nil {
+			return errors.Wrap(err, "deduplication based on .crc32 content")
+		}
+
+		if err := w.basedOnCrc32Content(ctx, d, h, dataDst, ssTablesGroupByID); err != nil {
+			return errors.Wrap(err, "deduplication based on .crc32 content")
+		}
+
+		return nil
+	}
+
+	notify := func(i int, err error) {
+		d := dirs[i]
+		w.Logger.Error(ctx, "Failed to deduplicate host",
+			"host", d.Host,
+			"keyspace", d.Keyspace,
+			"table", d.Table,
+			"error", err,
+		)
+	}
+
+	return parallel.Run(len(dirs), 1, f, notify)
+}
+
+func (w *worker) basedOnUUIDGenerationAvailability(ctx context.Context, d snapshotDir, h hostInfo,
+	remoteFiles []string, ssTablesGroupByID map[string][]string,
+) error {
+	// Per every SSTable files group, check if the name uses UUID to identify the SSTable generation.
+	// If the above is true, then check if files exists in the remote storage and remove it locally if exists in rmeote.
+	for id, ssTableContent := range ssTablesGroupByID {
+		// Check if id is an integer (no UUID)
+		if _, err := strconv.Atoi(id); err == nil {
+			continue
+		}
+
+		// Check if SSTable content is already available in remote
+		for _, file := range ssTableContent {
+			if slices.Contains(remoteFiles, file) {
+				// remove if exists in remote
+				localSSTableFileNameWithPath := path.Join(d.Path, file)
+				w.Logger.Info(ctx, "Removing local snapshot file (deduplication based on generation UUID)",
+					"host", h.IP, "file", localSSTableFileNameWithPath)
+				if err := w.Client.RcloneDeleteFile(ctx, h.IP, localSSTableFileNameWithPath); err != nil {
+					return errors.Wrapf(err, "cannot delete local snapshot's SSTable file %s", localSSTableFileNameWithPath)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (w *worker) basedOnCrc32Content(ctx context.Context, d snapshotDir, h hostInfo, dataDst string,
+	ssTablesGroupByID map[string][]string,
+) error {
+	// Reference to SSTables 3.0 Data File Format
+	// https://opensource.docs.scylladb.com/stable/architecture/sstable/sstable3/sstables-3-data-file-format.html
+
+	// Per every SSTable files group, compare local <ID>-Digest.crc32 content
+	// to the remote <ID>-Digest.crc32 content.
+	// The same content implies that SSTable can be deduplicated and removed from local directory.
+	for _, ssTableContent := range ssTablesGroupByID {
+		crc32Idx := slices.IndexFunc(ssTableContent, func(s string) bool {
+			return strings.HasSuffix(s, "Digest.crc32")
+		})
+		if crc32Idx == -1 {
+			continue
+		}
+
+		remoteCRC32FileNameWithPath := path.Join(dataDst, ssTableContent[crc32Idx])
+		remoteCRC32, err := w.Client.RcloneCat(ctx, h.IP, remoteCRC32FileNameWithPath)
+		if err != nil {
+			if strings.Contains(err.Error(), "object not found") {
+				continue
+			}
```

Comment:
I changed it a bit to not send map of ID->list of files for remote, but just set of SSTable ID on remote.
Let me keep this "object not found" check, but just validate if SSTable of given generation ID exists on remote.

Reply:
But still, this should never happen and if it does, it means that something is wrong - I think it's safer to return error in such case.
Also, string error checking is problematic in general - it would be better to check for http status code.

So what is the motivation behind this check?
Are you worried that there might be some really old sstables which don't have the .crc32 file?
But if that's the case, we should just change the map key from sstable ID to a pair of sstable ID and format.

Reply:
makes sense to check it against HTTP error code (404)
If you ask about the motivation, then I want to avoid hard stop on backup whenever the file just doesn't exist. Although I think it's not probable, as the code is looking for the file in sstablecontent.

Reply:
Ok, but please use `http.StatusNotFound` instead of raw 404 int.